lib/api/CDataFrameTrainBoostedTreeClassifierRunner.cc

```diff
@@ -182,14 +187,44 @@ void CDataFrameTrainBoostedTreeClassifierRunner::writeOneRow(
                                 writer.Key(classValues[j]);
                                 writer.Double(shap[i](j));
                             }
-                            writer.Key(CDataFrameTrainBoostedTreeRunner::IMPORTANCE_FIELD_NAME);
+                            writer.Key(IMPORTANCE_FIELD_NAME);
                             writer.Double(shap[i].lpNorm<1>());
                         }
                         writer.EndObject();
                     }
                 }
                 writer.EndArray();
+
+                for (std::size_t i = 0; i < shap.size(); ++i) {
```

Comment:
This callback only gets passed the top values whereas really it should get passed all to make these accurate. I wonder if this could be rejigged slightly so we get all the values.

Reply:
The relevant part from `CTreeShapFeatureImportance.cc` is 
```

    m_TopShapValues.resize(m_ReducedShapValues.size());
    std::iota(m_TopShapValues.begin(), m_TopShapValues.end(), 0);
    if (m_NumberTopShapValues < m_TopShapValues.size()) {
        std::nth_element(m_TopShapValues.begin(), m_TopShapValues.begin() + m_NumberTopShapValues,
                         m_TopShapValues.end(), [this](std::size_t lhs, std::size_t rhs) {
                             return COrderings::lexicographical_compare(
                                 -las::L1(m_ReducedShapValues[lhs]), lhs,
                                 -las::L1(m_ReducedShapValues[rhs]), rhs);
                         });
        m_TopShapValues.resize(m_NumberTopShapValues);
        std::sort(m_TopShapValues.begin(), m_TopShapValues.end());
    }

    writer(m_TopShapValues, m_ColumnNames, m_ReducedShapValues);
```

The callback works on `m_ReducedShapValues` which AFAICS still contains all shap values.

Reply:
Ok cool. I wonder if I'd remembered an earlier version in this comment. Anyway, as you observe it is the indices which are reduced and not the actual values.